collector/nvidia.go

```diff
@@ -0,0 +1,305 @@
+	device      C.nvmlDevice_t
+	driverModel C.nvmlDriverModel_t
+	name        string
+}
+
+type nvidiaCollector struct {
+	TotalProcessGpuMemoryUsed *prometheus.Desc
+	TotalGpuMemoryUsed        *prometheus.Desc
+
+	gpuProcessWhitelistPattern *regexp.Regexp
+	gpuProcessBlacklistPattern *regexp.Regexp
+
+	nvmlLibrary C.HINSTANCE
+
+	devices []nvdiaGpuDevice
+}
+
+func (error C.nvmlReturn_t) Error() string {
+	return C.GoString(C.invoke_nvmlErrorString(error))
+}
+
+func clen(bytes []byte) int {
+	for i := 0; i < len(bytes); i++ {
+		if bytes[i] == 0 {
+			return i
+		}
+	}
+	return 0
+}
+
+// newNvidiaCollector ...
+func newNvidiaCollector() (Collector, error) {
+	nvmlLibrary := C.LoadLibrary(C.CString("nvml.dll"))
+	if nvmlLibrary == C.HINSTANCE(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlInit_v2Ptr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlInit_v2"))
+	if C.nvmlInit_v2Ptr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlErrorStringPtr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlErrorString"))
+	if C.nvmlErrorStringPtr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetCount_v2Ptr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetCount_v2"))
+	if C.nvmlDeviceGetCount_v2Ptr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetHandleByIndex_v2Ptr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetHandleByIndex_v2"))
+	if C.nvmlDeviceGetHandleByIndex_v2Ptr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetNamePtr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetName"))
+	if C.nvmlDeviceGetNamePtr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetGraphicsRunningProcesses_v2Ptr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetGraphicsRunningProcesses_v2"))
+	if C.nvmlDeviceGetGraphicsRunningProcesses_v2Ptr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlSystemGetProcessNamePtr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlSystemGetProcessName"))
+	if C.nvmlSystemGetProcessNamePtr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetDriverModelPtr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetDriverModel"))
+	if C.nvmlDeviceGetDriverModelPtr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetMemoryInfoPtr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetMemoryInfo"))
+	if C.nvmlDeviceGetMemoryInfoPtr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	err := C.invoke_nvmlInit_v2()
+	if err != C.NVML_SUCCESS {
+		return nil, err
+	}
+
+	const subsystem = "nvidia"
+
+	if *processWhitelist == ".*" && *processBlacklist == "" {
+		log.Warn("No filters specified for nvidia collector. This will generate a very large number of metrics!")
```

Comment:
Is this a copy/paste, or will there actually be a lot of metrics?

Reply:
That is not a copy/paste, since my initial thought was that basically any process that is using the GPU (thus for minimum any process that has a GUI) will be exposed here - ofc there won't be as many per process metrics as in the process collector, but the actual entries will be as many as there are processes using the GPU. Again, something we can discuss and work out.

Reply:
Okay, then no worries! It's just been a rather frequent occurrence historically that contributors start off by copy/pasting one of the collects that generate a lot of metrics and gets this log line by chance, so to speak. From your description it sounds warranted then, I wasn't sure how many matches it would result in.